File: services/rust/rocket/src/routes/handlers/put.rs
use crate::routes::{
    errors::AppError,
    extractors::{AppJson, AppPath},
    guards::ChaosGuard,
};
use app::{state::AppState, task::Task};
use rocket::{State, http::Status, put, serde::json::Json};
use serde_json::{Value, json};
use std::sync::{Arc, Mutex};
use tracing::{error, info, instrument, warn};

#[put("/<id>", data = "<body>")]
#[instrument(skip_all)]
pub async fn put_handler(
    _guard: ChaosGuard,
    id: AppPath,
    state: &State<Arc<Mutex<AppState>>>,
    body: AppJson<Task>,
) -> (Status, Json<Value>) {
    let id = id.into_inner();

    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let previous_task = task.clone();
            *task = body.into_inner();

            info!(
                %id,
                from_secret = previous_task.secret.len(),
                to_secret = task.secret.len(),
                from_operation = previous_task.operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PUT",
                "Overwrote task"
            );

            return (Status::Ok, Json(json!(task.to_owned())));
        } else {
            drop(state);
            warn!(%id, method = "PUT", "Task not found");
            return AppError::TaskNotFound.into_response();
        }
    }

    error!(%id, method = "PUT", "Poisoned lock");

    AppError::Internal.into_response()
}

File: services/go/fiber/routes/handlers/put.go
package handlers

import (
	"app"
	"fiber/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gofiber/fiber/v3"
)

func PutHandler(state *app.AppState) fiber.Handler {
	return func(c fiber.Ctx) error {
		id, err := ParseID(c)
		if err != nil {
			return err
		}

		newTask, err := ParseTask(c, c.Body())
		if err != nil {
			return err
		}

		state.Mu.Lock()
		previousTask, ok := state.Tasks[id]
		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", c.Method(), "path", c.Path(), "id", id.String())
			return routes.AppErrors.TaskNotFound.Error(c)
		}
		state.Tasks[id] = newTask
		state.Mu.Unlock()

		slog.Info(
			"Overwrote task",
			"id", id.String(),
			"from_secret", len(previousTask.Secret),
			"to_secret", len(newTask.Secret),
			"from_operation", strings.ToLower(string(previousTask.Operation)),
			"to_operation", strings.ToLower(string(newTask.Operation)),
			"method", c.Method(),
		)

		return c.Status(http.StatusOK).JSON(newTask)
	}
}
